src/Middleware/OutputCaching/src/OutputCacheMiddleware.cs

```diff
@@ -341,32 +341,33 @@ internal void CreateCacheKey(OutputCacheContext context)
             return;
         }
 
-        var varyHeaders = context.CacheVaryByRules.Headers;
+        var varyHeaderNames = context.CacheVaryByRules.HeaderNames;
+        var varyRouteValueNames = context.CacheVaryByRules.RouteValueNames;
         var varyQueryKeys = context.CacheVaryByRules.QueryKeys;
-        var varyByCustomKeys = context.CacheVaryByRules.VaryByCustom;
+        var varyByCustomKeys = context.CacheVaryByRules.HasVaryByCustom ? context.CacheVaryByRules.VaryByCustom : null;
         var varyByPrefix = context.CacheVaryByRules.VaryByPrefix;
 
         // Check if any vary rules exist
-        if (!StringValues.IsNullOrEmpty(varyHeaders) || !StringValues.IsNullOrEmpty(varyQueryKeys) || !StringValues.IsNullOrEmpty(varyByPrefix) || varyByCustomKeys?.Count > 0)
+        if (!StringValues.IsNullOrEmpty(varyHeaderNames) || !StringValues.IsNullOrEmpty(varyRouteValueNames) || !StringValues.IsNullOrEmpty(varyQueryKeys) || !StringValues.IsNullOrEmpty(varyByPrefix) || varyByCustomKeys?.Count > 0)
```

Comment:
This line is pretty long, may be better to wrap it

Reply:
I have another PR that will remove it, so I will ignore it for now (if I may).